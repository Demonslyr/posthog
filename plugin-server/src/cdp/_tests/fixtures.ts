import { randomUUID } from 'crypto'
import { Message } from 'node-rdkafka'

import { insertRow } from '~/tests/helpers/sql'

import { ClickHouseTimestamp, ProjectId, RawClickHouseEvent, Team } from '../../types'
import { PostgresRouter } from '../../utils/db/postgres'
import { UUIDT } from '../../utils/utils'
import { CdpInternalEvent } from '../schema'
import { compileHog } from '../templates/compiler'
import {
    HogFunctionInvocation,
    HogFunctionInvocationGlobals,
    HogFunctionInvocationGlobalsWithInputs,
    HogFunctionTemplateType,
    HogFunctionType,
    IntegrationType,
} from '../types'

export const createHogFunction = (hogFunction: Partial<HogFunctionType>) => {
    const item: HogFunctionType = {
        id: randomUUID(),
        type: 'destination',
        name: 'Hog Function',
        team_id: 1,
        enabled: true,
        ...hogFunction,
    } as HogFunctionType

    return item
}

export const createIntegration = (integration: Partial<IntegrationType>) => {
    const item: IntegrationType = {
        team_id: 1,
        errors: '',
        created_at: new Date().toISOString(),
        created_by_id: 1001,
        id: integration.id ?? 1,
        kind: integration.kind ?? 'slack',
        config: {},
        sensitive_config: {},
        ...integration,
    }

    return item
}

export const createIncomingEvent = (teamId: number, data: Partial<RawClickHouseEvent>): RawClickHouseEvent => {
    return {
        team_id: teamId,
        project_id: teamId as ProjectId,
        created_at: new Date().toISOString() as ClickHouseTimestamp,
        elements_chain: '[]',
        person_created_at: new Date().toISOString() as ClickHouseTimestamp,
        person_properties: '{}',
        distinct_id: 'distinct_id_1',
        uuid: randomUUID(),
        event: '$pageview',
        timestamp: new Date().toISOString() as ClickHouseTimestamp,
        properties: '{}',
        person_mode: 'full',
        ...data,
    }
}

export const createKafkaMessage = (event: any, overrides: Partial<Message> = {}): Message => {
    return {
        partition: 1,
        topic: 'test',
        offset: 0,
        timestamp: overrides.timestamp ?? Date.now(),
        size: 1,
        ...overrides,
        value: Buffer.from(JSON.stringify(event)),
    }
}

export const createInternalEvent = (teamId: number, data: Partial<CdpInternalEvent>): CdpInternalEvent => {
    return {
        team_id: teamId,
        event: {
            timestamp: new Date().toISOString(),
            properties: {},
            uuid: randomUUID(),
            event: '$pageview',
            distinct_id: 'distinct_id',
        },
        ...data,
    }
}

export const insertHogFunction = async (
    postgres: PostgresRouter,
    team_id: Team['id'],
    hogFunction: Partial<HogFunctionType> = {}
): Promise<HogFunctionType> => {
    // This is only used for testing so we need to override some values
    const res = await insertRow(postgres, 'posthog_hogfunction', {
        ...createHogFunction({
            ...hogFunction,
            team_id: team_id,
        }),
        // Ensure these fields are null unless explicitly set
        hog: hogFunction.hog ?? null,
        bytecode: hogFunction.bytecode ?? null,
        inputs_schema: hogFunction.inputs_schema ?? null,
        mappings: hogFunction.mappings ?? null,
        description: '',
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
        created_by_id: 1001,
        deleted: false,
    })
    return res
}

export const insertIntegration = async (
    postgres: PostgresRouter,
    team_id: Team['id'],
    integration: Partial<IntegrationType> = {}
): Promise<IntegrationType> => {
    const res = await insertRow(
        postgres,
        'posthog_integration',
        createIntegration({
            ...integration,
            team_id: team_id,
        })
    )
    return res
}

export const createHogExecutionGlobals = (
    data: Partial<HogFunctionInvocationGlobals> = {}
): HogFunctionInvocationGlobals => {
    return {
        groups: {},
        ...data,
        person: {
            id: 'uuid',
            name: 'test',
            url: 'http://localhost:8000/persons/1',
            properties: {
                email: 'test@posthog.com',
                first_name: 'Pumpkin',
            },
            ...(data.person ?? {}),
        },
        project: {
            id: 1,
            name: 'test',
            url: 'http://localhost:8000/projects/1',
            ...(data.project ?? {}),
        },
        event: {
            uuid: 'uuid',
            event: 'test',
            elements_chain: '',
            distinct_id: 'distinct_id',
            url: 'http://localhost:8000/events/1',
            properties: {
                $lib_version: '1.2.3',
            },
            timestamp: new Date().toISOString(),
            ...(data.event ?? {}),
        },
    }
}

export const createInvocation = (
    _hogFunction: Partial<HogFunctionType> = {},
    _globals: Partial<HogFunctionInvocationGlobals> = {}
): HogFunctionInvocation => {
    const hogFunction = createHogFunction(_hogFunction)
    // Add the source of the trigger to the globals

    const globals = createHogExecutionGlobals(_globals)
    globals.source = {
        name: hogFunction.name ?? `Hog function: ${hogFunction.id}`,
        url: `${globals.project.url}/pipeline/destinations/hog-${hogFunction.id}/configuration/`,
    }

    return {
        id: new UUIDT().toString(),
        // NOTE: This is due to some legacy code that checks for inputs and uses it. BW will fix later.
        globals: globals as HogFunctionInvocationGlobalsWithInputs,
        teamId: hogFunction.team_id,
        hogFunction,
        queue: 'hog',
        timings: [],
        queuePriority: 0,
    }
}

export const createHogFunctionTemplate = async (template: Partial<HogFunctionTemplateType>) => ({
    id: randomUUID(),
    template_id: 'default_template_id',
    code: 'return event',
    bytecode: await compileHog('return event'),
    inputs_schema: [
        {
            type: 'string',
            key: 'test_input',
            label: 'Test Input',
            required: true,
            default: 'test',
        },
        {
            type: 'boolean',
            key: 'enabled',
            label: 'Enabled',
            default: true,
        },
    ],
    mappings: [
        {
            inputs_schema: [
                {
                    type: 'string',
                    key: 'field_mapping',
                    label: 'Field Mapping',
                    required: true,
                    default: 'event.properties',
                },
            ],
            inputs: {
                field_mapping: {
                    value: 'event.properties',
                },
            },
            filters: null,
        },
    ],
    sha: 'sha1',
    name: 'Test Template',
    code_language: 'hog',
    type: 'destination',
    status: 'alpha',
    category: '[]',
    free: false,
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
    ...template,
})

export const insertHogFunctionTemplate = async (
    postgres: PostgresRouter,
    template: Partial<HogFunctionTemplateType> = {}
): Promise<HogFunctionTemplateType> => {
    const res = await insertRow(postgres, 'posthog_hogfunctiontemplate', {
        ...(await createHogFunctionTemplate({
            ...template,
        })),
    })
    return res
}
