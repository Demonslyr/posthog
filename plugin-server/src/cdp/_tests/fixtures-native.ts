import { UUIDT } from '~/utils/utils'

import { CyclotronJobInvocationHogFunction, HogFunctionType } from '../types'
import { createHogFunction } from './fixtures'
import { SAMPLE_GLOBALS } from './fixtures'

export const createExampleNativeInvocation = (
    _hogFunction: Partial<HogFunctionType> = {},
    inputs: Record<string, any> = {}
): CyclotronJobInvocationHogFunction => {
    const hogFunction = createHogFunction(_hogFunction)

    return {
        id: new UUIDT().toString(),
        state: {
            globals: {
                inputs,
                ...SAMPLE_GLOBALS,
            },
            timings: [],
            attempts: 0,
        },
        teamId: hogFunction.team_id,
        functionId: hogFunction.id,
        hogFunction,
        queue: 'native',
        queuePriority: 0,
    }
}
