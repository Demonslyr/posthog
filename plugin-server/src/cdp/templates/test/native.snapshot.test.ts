import { DateTime, Settings } from 'luxon'

import { SAMPLE_GLOBALS } from '~/cdp/_tests/fixtures'

import { NATIVE_HOG_FUNCTIONS } from '../index'
import { DestinationTester, generateTestData } from './test-helpers'

for (const template of NATIVE_HOG_FUNCTIONS) {
    const testDestination = new DestinationTester(template)

    describe(`Testing snapshots for ${template.id} destination:`, () => {
        beforeEach(() => {
            Settings.defaultZone = 'UTC'
            const fixedTime = DateTime.fromObject({ year: 2025, month: 1, day: 1 }, { zone: 'UTC' })
            jest.spyOn(Date, 'now').mockReturnValue(fixedTime.toMillis())
        })

        afterEach(() => {
            Settings.defaultZone = 'system'
            jest.useRealTimers()
        })

        for (const mapping of template.mapping_templates) {
            it.each(['required fields', 'all fields'])(`${mapping.name} mapping - %s`, async (required) => {
                const seedName = `${template.id}#${mapping.name}`
                const inputs = generateTestData(seedName, template.inputs_schema, required === 'required fields')
                const mappingInputs = generateTestData(
                    seedName,
                    mapping.inputs_schema ?? [],
                    required === 'required fields'
                )

                const responses = await testDestination.invokeMapping(
                    mapping.name,
                    SAMPLE_GLOBALS,
                    { ...inputs, debug_mode: true },
                    mappingInputs
                )

                responses.logs.forEach((x) => {
                    if (typeof x.message === 'string' && x.message.includes('Function completed in')) {
                        x.message = 'Function completed in [REPLACED]'
                    }
                })
                responses.invocation.id = 'invocation-id'

                expect(responses).toMatchSnapshot()
            })
        }
    })
}
