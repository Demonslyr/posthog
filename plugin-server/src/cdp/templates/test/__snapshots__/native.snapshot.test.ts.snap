// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing snapshots for native-posthog destination: Send an event mapping - all fields 1`] = `
{
  "capturedPostHogEvents": [],
  "finished": true,
  "invocation": {
    "functionId": "native-posthog",
    "hogFunction": {
      "actions": {
        "event": {
          "perform": [Function],
        },
      },
      "bytecode": [],
      "category": [
        "Analytics",
      ],
      "created_at": "2024-01-01T00:00:00Z",
      "deleted": false,
      "description": "Sends events to PostHog",
      "enabled": true,
      "free": false,
      "hog": "return event",
      "icon_url": "/static/posthog-icon.svg",
      "id": "native-posthog",
      "inputs": {
        "apiKey": {
          "bytecode": [
            "_H",
            1,
            32,
            "9OSA[^iPUw",
            38,
          ],
          "value": "9OSA[^iPUw",
        },
        "debug_mode": {
          "bytecode": undefined,
          "value": true,
        },
      },
      "inputs_schema": [
        {
          "description": "API Key to authenticate the request.",
          "key": "apiKey",
          "label": "API Key",
          "required": true,
          "secret": false,
          "type": "string",
        },
        {
          "default": false,
          "description": "Will log configuration and request details",
          "key": "debug_mode",
          "label": "Debug Mode",
          "type": "boolean",
        },
      ],
      "is_addon_required": false,
      "mapping_templates": [
        {
          "associated_action": "event",
          "filters": {
            "events": [
              {
                "id": null,
                "name": "All events",
                "order": 0,
                "type": "events",
              },
            ],
          },
          "include_by_default": true,
          "inputs_schema": [
            {
              "description": "The name of the event to send.",
              "key": "eventName",
              "label": "Event Name",
              "required": true,
              "secret": false,
              "type": "string",
            },
            {
              "description": "The ID of the event to send.",
              "key": "eventId",
              "label": "Event ID",
              "required": false,
              "secret": false,
              "type": "string",
            },
            {
              "default": {
                "$$_extend_object": "{event.properties}",
              },
              "description": "Properties to send with the event.",
              "key": "eventProperties",
              "label": "Event Properties",
              "required": false,
              "secret": false,
              "type": "dictionary",
            },
            {
              "default": "event",
              "description": "The associated mapping to use",
              "hidden": true,
              "key": "internal_associated_mapping",
              "label": "Associated Mapping",
              "required": true,
              "secret": false,
              "type": "string",
            },
          ],
          "name": "Send an event",
        },
      ],
      "mappings": [
        {
          "associated_action": "event",
          "filters": {
            "events": [
              {
                "id": null,
                "name": "All events",
                "order": 0,
                "type": "events",
              },
            ],
          },
          "include_by_default": true,
          "inputs": {
            "eventId": {
              "bytecode": [
                "_H",
                1,
                32,
                "9OSA[^iPUw",
                38,
              ],
              "value": "9OSA[^iPUw",
            },
            "eventName": {
              "bytecode": [
                "_H",
                1,
                32,
                "9OSA[^iPUw",
                38,
              ],
              "value": "9OSA[^iPUw",
            },
            "eventProperties": {
              "bytecode": {
                "$$_extend_object": [
                  "_H",
                  1,
                  32,
                  "properties",
                  32,
                  "event",
                  1,
                  2,
                  38,
                ],
              },
              "value": {
                "$$_extend_object": "{event.properties}",
              },
            },
            "internal_associated_mapping": {
              "bytecode": [
                "_H",
                1,
                32,
                "event",
                38,
              ],
              "value": "event",
            },
          },
          "inputs_schema": [
            {
              "description": "The name of the event to send.",
              "key": "eventName",
              "label": "Event Name",
              "required": true,
              "secret": false,
              "type": "string",
            },
            {
              "description": "The ID of the event to send.",
              "key": "eventId",
              "label": "Event ID",
              "required": false,
              "secret": false,
              "type": "string",
            },
            {
              "default": {
                "$$_extend_object": "{event.properties}",
              },
              "description": "Properties to send with the event.",
              "key": "eventProperties",
              "label": "Event Properties",
              "required": false,
              "secret": false,
              "type": "dictionary",
            },
            {
              "default": "event",
              "description": "The associated mapping to use",
              "hidden": true,
              "key": "internal_associated_mapping",
              "label": "Associated Mapping",
              "required": true,
              "secret": false,
              "type": "string",
            },
          ],
          "name": "Send an event",
        },
      ],
      "name": "PostHog",
      "status": "hidden",
      "team_id": 1,
      "template_id": "native-posthog",
      "type": "destination",
      "updated_at": "2024-01-01T00:00:00Z",
    },
    "id": "invocation-id",
    "queue": "native",
    "queueMetadata": {
      "tries": 1,
    },
    "queueParameters": undefined,
    "queuePriority": 0,
    "queueScheduledAt": undefined,
    "queueSource": undefined,
    "state": {
      "attempts": 0,
      "globals": {
        "event": {
          "distinct_id": "distinct_id",
          "elements_chain": "",
          "event": "test",
          "properties": {
            "email": "test@posthog.com",
          },
          "timestamp": "",
          "url": "",
          "uuid": "uuid",
        },
        "inputs": {
          "apiKey": "9OSA[^iPUw",
          "debug_mode": true,
          "eventId": "9OSA[^iPUw",
          "eventName": "9OSA[^iPUw",
          "eventProperties": {
            "email": "test@posthog.com",
          },
          "internal_associated_mapping": "event",
        },
        "person": {
          "id": "person-id",
          "name": "person-name",
          "properties": {
            "email": "example@posthog.com",
          },
          "url": "https://us.posthog.com/projects/1/persons/1234",
        },
        "project": {
          "id": 1,
          "name": "project-name",
          "url": "https://us.posthog.com/projects/1",
        },
        "source": {
          "name": "hog-function-name",
          "url": "https://us.posthog.com/hog_functions/1234",
        },
      },
      "timings": [],
    },
    "teamId": 1,
  },
  "logs": [
    {
      "level": "debug",
      "message": "config, {"debug_mode":true,"apiKey":"9OSA[^iPUw","eventName":"9OSA[^iPUw","eventId":"9OSA[^iPUw","eventProperties":{"email":"test@posthog.com"},"internal_associated_mapping":"event"}",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "debug",
      "message": "endpoint, http://localhost:2080/7c138c0e-e208-4bc0-8378-4bbbdedad5bf",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "debug",
      "message": "options, {"method":"POST","headers":{"Content-Type":"application/json","Authorization":"Bearer 9OSA[^iPUw"},"json":{"event":"9OSA[^iPUw","eventId":"9OSA[^iPUw","properties":{"email":"test@posthog.com"}}}",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "debug",
      "message": "fetchOptions, {"method":"POST","headers":{"User-Agent":"PostHog.com/1.0","Content-Type":"application/json","Authorization":"Bearer 9OSA[^iPUw"},"body":"{\\"event\\":\\"9OSA[^iPUw\\",\\"eventId\\":\\"9OSA[^iPUw\\",\\"properties\\":{\\"email\\":\\"test@posthog.com\\"}}"}",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "debug",
      "message": "convertedResponse, 200, {"status":"OK"}, {"status":"OK"}, {"content-type":"application/json"}",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "info",
      "message": "Function completed in [REPLACED]",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
  ],
  "metrics": [],
}
`;

exports[`Testing snapshots for native-posthog destination: Send an event mapping - required fields 1`] = `
{
  "capturedPostHogEvents": [],
  "finished": true,
  "invocation": {
    "functionId": "native-posthog",
    "hogFunction": {
      "actions": {
        "event": {
          "perform": [Function],
        },
      },
      "bytecode": [],
      "category": [
        "Analytics",
      ],
      "created_at": "2024-01-01T00:00:00Z",
      "deleted": false,
      "description": "Sends events to PostHog",
      "enabled": true,
      "free": false,
      "hog": "return event",
      "icon_url": "/static/posthog-icon.svg",
      "id": "native-posthog",
      "inputs": {
        "apiKey": {
          "bytecode": [
            "_H",
            1,
            32,
            "9OSA[^iPUw",
            38,
          ],
          "value": "9OSA[^iPUw",
        },
        "debug_mode": {
          "bytecode": undefined,
          "value": true,
        },
      },
      "inputs_schema": [
        {
          "description": "API Key to authenticate the request.",
          "key": "apiKey",
          "label": "API Key",
          "required": true,
          "secret": false,
          "type": "string",
        },
        {
          "default": false,
          "description": "Will log configuration and request details",
          "key": "debug_mode",
          "label": "Debug Mode",
          "type": "boolean",
        },
      ],
      "is_addon_required": false,
      "mapping_templates": [
        {
          "associated_action": "event",
          "filters": {
            "events": [
              {
                "id": null,
                "name": "All events",
                "order": 0,
                "type": "events",
              },
            ],
          },
          "include_by_default": true,
          "inputs_schema": [
            {
              "description": "The name of the event to send.",
              "key": "eventName",
              "label": "Event Name",
              "required": true,
              "secret": false,
              "type": "string",
            },
            {
              "description": "The ID of the event to send.",
              "key": "eventId",
              "label": "Event ID",
              "required": false,
              "secret": false,
              "type": "string",
            },
            {
              "default": {
                "$$_extend_object": "{event.properties}",
              },
              "description": "Properties to send with the event.",
              "key": "eventProperties",
              "label": "Event Properties",
              "required": false,
              "secret": false,
              "type": "dictionary",
            },
            {
              "default": "event",
              "description": "The associated mapping to use",
              "hidden": true,
              "key": "internal_associated_mapping",
              "label": "Associated Mapping",
              "required": true,
              "secret": false,
              "type": "string",
            },
          ],
          "name": "Send an event",
        },
      ],
      "mappings": [
        {
          "associated_action": "event",
          "filters": {
            "events": [
              {
                "id": null,
                "name": "All events",
                "order": 0,
                "type": "events",
              },
            ],
          },
          "include_by_default": true,
          "inputs": {
            "eventName": {
              "bytecode": [
                "_H",
                1,
                32,
                "9OSA[^iPUw",
                38,
              ],
              "value": "9OSA[^iPUw",
            },
            "eventProperties": {
              "bytecode": {
                "$$_extend_object": [
                  "_H",
                  1,
                  32,
                  "properties",
                  32,
                  "event",
                  1,
                  2,
                  38,
                ],
              },
              "value": {
                "$$_extend_object": "{event.properties}",
              },
            },
            "internal_associated_mapping": {
              "bytecode": [
                "_H",
                1,
                32,
                "event",
                38,
              ],
              "value": "event",
            },
          },
          "inputs_schema": [
            {
              "description": "The name of the event to send.",
              "key": "eventName",
              "label": "Event Name",
              "required": true,
              "secret": false,
              "type": "string",
            },
            {
              "description": "The ID of the event to send.",
              "key": "eventId",
              "label": "Event ID",
              "required": false,
              "secret": false,
              "type": "string",
            },
            {
              "default": {
                "$$_extend_object": "{event.properties}",
              },
              "description": "Properties to send with the event.",
              "key": "eventProperties",
              "label": "Event Properties",
              "required": false,
              "secret": false,
              "type": "dictionary",
            },
            {
              "default": "event",
              "description": "The associated mapping to use",
              "hidden": true,
              "key": "internal_associated_mapping",
              "label": "Associated Mapping",
              "required": true,
              "secret": false,
              "type": "string",
            },
          ],
          "name": "Send an event",
        },
      ],
      "name": "PostHog",
      "status": "hidden",
      "team_id": 1,
      "template_id": "native-posthog",
      "type": "destination",
      "updated_at": "2024-01-01T00:00:00Z",
    },
    "id": "invocation-id",
    "queue": "native",
    "queueMetadata": {
      "tries": 1,
    },
    "queueParameters": undefined,
    "queuePriority": 0,
    "queueScheduledAt": undefined,
    "queueSource": undefined,
    "state": {
      "attempts": 0,
      "globals": {
        "event": {
          "distinct_id": "distinct_id",
          "elements_chain": "",
          "event": "test",
          "properties": {
            "email": "test@posthog.com",
          },
          "timestamp": "",
          "url": "",
          "uuid": "uuid",
        },
        "inputs": {
          "apiKey": "9OSA[^iPUw",
          "debug_mode": true,
          "eventName": "9OSA[^iPUw",
          "eventProperties": {
            "email": "test@posthog.com",
          },
          "internal_associated_mapping": "event",
        },
        "person": {
          "id": "person-id",
          "name": "person-name",
          "properties": {
            "email": "example@posthog.com",
          },
          "url": "https://us.posthog.com/projects/1/persons/1234",
        },
        "project": {
          "id": 1,
          "name": "project-name",
          "url": "https://us.posthog.com/projects/1",
        },
        "source": {
          "name": "hog-function-name",
          "url": "https://us.posthog.com/hog_functions/1234",
        },
      },
      "timings": [],
    },
    "teamId": 1,
  },
  "logs": [
    {
      "level": "debug",
      "message": "config, {"debug_mode":true,"apiKey":"9OSA[^iPUw","eventName":"9OSA[^iPUw","eventProperties":{"email":"test@posthog.com"},"internal_associated_mapping":"event"}",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "debug",
      "message": "endpoint, http://localhost:2080/7c138c0e-e208-4bc0-8378-4bbbdedad5bf",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "debug",
      "message": "options, {"method":"POST","headers":{"Content-Type":"application/json","Authorization":"Bearer 9OSA[^iPUw"},"json":{"event":"9OSA[^iPUw","properties":{"email":"test@posthog.com"}}}",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "debug",
      "message": "fetchOptions, {"method":"POST","headers":{"User-Agent":"PostHog.com/1.0","Content-Type":"application/json","Authorization":"Bearer 9OSA[^iPUw"},"body":"{\\"event\\":\\"9OSA[^iPUw\\",\\"properties\\":{\\"email\\":\\"test@posthog.com\\"}}"}",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "debug",
      "message": "convertedResponse, 200, {"status":"OK"}, {"status":"OK"}, {"content-type":"application/json"}",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
    {
      "level": "info",
      "message": "Function completed in [REPLACED]",
      "timestamp": "2025-01-01T01:00:00.000+01:00",
    },
  ],
  "metrics": [],
}
`;
